### DISCUSSION TITLE
Escape sequences in windows

### DISCUSSION THREAD

**Discussion by hubertushirsch** [ole00/afterburner#62] [General]
The Afterburner PC program uses ANSI escape sequences on some outputs.
This does not always work depending on the Windows version and the console versions (CMD or Powershell).
If you see the following output:
```
D:\<path> .\afterburner_w64.exe -d COM15 -t ATF16V8B v -f testfile.jed
Uploading fuse map...
2194/2194 |########################################|     |
←[?25hvarSetVppIndex 0
varSetVpp 3:130
....
```
the escape sequences are not recognized and are printed as normal text..
(Additional vertical bar at the end of the **###** line and the string **←[?25h**)

Then try to set the following registry entry:
```
Windows Registry Editor Version 5.00
[HKEY_CURRENT_USER\Console]
"VirtualTerminalLevel"=dword:00000001
```
Afterwards the output should look like this:
```
D:\<path> .\afterburner_w64.exe -d COM15 -t ATF16V8B v -f testfile.jed
Uploading fuse map...
2194/2194 |########################################|
varSetVppIndex 0
varSetVpp 3:130
....
```
If setting this registry entry has no effect, you can alternatively modify the PC program (afterburner.c) and omit the escape sequences when compiling for Windows.
```
#ifndef _USE_WIN_API_
     printf("\e[?25l");
#endif
....
#ifndef _USE_WIN_API_
     printf("\e[?25h");
#endif
```

**Comment by hubertushirsch** [comment]
You can use this file to import the registry key.
To do this, you have to change the extension of the attached file from *.txt to *.reg.
Or you can enter the key/value manually in the Windows registry editor.

[Enable_Windows_Console_ESC_Sequences.txt](https://github.com/ole00/afterburner/files/14728593/Enable_Windows_Console_ESC_Sequences.txt)

**Comment by ole00** [comment]
Thanks for the ideas. I did not use windows version, but I think using regular '\r' shout reposition cursor to the beginning of the line, without a line feed, so perhaps that should be the simple fix for the progress bar. I'll take a look.
EDIT: I see, that  escape sequence is there to hide cursor. and '\r' was there already. Without the escape sequence the progress bar was a bit weird (jumpy), but when adding fflush(stdout); just after the printf the progress bar animation looks OK and the cursor does not cause weird artefacts.

**Comment by ole00** [comment]
It should be now fixed in this commit a48de726bf71c9f8af55a469b738fd8eb37d3533
I tested it on Win7, CMD shell.  The PC app release is also updated.

**Comment by hubertushirsch** [comment]
I didn't expect you to change the code.
I just wanted to give Windows users a tip about which settings the Afterburner PC program works best with.
That's why I didn't post it as an issue, but as a discussion.
Anyway, thanks for the code change that fixes the output problem in a different way.

Hubert

**Comment by hubertushirsch** [comment]
Fix testet. Works.
Also testet the last change - serial port detection.

THX

**Comment by ole00** [comment]
Thanks for the feedback and testing.  Appreciated.
